### PR TITLE
feat(): faster transpilation

### PR DESCRIPTION
**Pull request by chyzwar**
This is early stage I just wanted to getr some feedback from you @kdy1 
I am speeding up dir compilation. 

Currently, for three.js
```
yarn swc src -d build --source-maps true --copy-filesyarn run v1.22.10

start 1622221738322
parseoptions: 1.218ms
Successfully compiled 374 files with swc.
Compilation dir1: 561.012ms
Done in 0.80s.
```

with dir2
```
yarn swc src -d build --source-maps true --copy-files

start 1622221821383
parseoptions: 1.087ms
Compilation: 289.231ms

  failed: 0
  compiled: 374
  copied: 0
  omitted: 0
  
Done in 0.55s.

```
Without source maps:
```
parseoptions: 1.157ms
Successfully compiled 374 files with swc.
Compilation dir1: 471.063ms
Done in 0.71s.
```

```
yarn swc src -d build --source-maps false

parseoptions: 1.162ms
Compilation: 223.219ms

  failed: 0
  compiled: 374
  copied: 0
  omitted: 0
  
Done in 0.47s.
```
It would be almost **50% faster** when compiling with dir2. In general process startup take almost as much as compilation, this might be local quirk with nodenv.

### CODE REVIEW

**Comment by chyzwar** [comment]
Another perf win could be handling of sourcemaps
```
    if (result.map) {
      const map = convertSourceMap.fromJSON(result.map);
      // TODO: fix this in core
      // https://github.com/swc-project/swc/issues/1388
      if (opts.sourceFileName) {
        map.getProperty('sources')[0] = opts.sourceFileName;
      }
      if (opts.sourceRoot) {
        map.setProperty('sourceRoot', opts.sourceRoot);
      }
      result.map = map.toJSON();
    }
```
This parse and stringify map to apply fix in source maps, Once this is fixed in core we should get 10-20milisec

**Comment by kdy1** [comment]
Wow, I'm a bit shocked by the perf gap.
I didn't know that the time is consumed by js code.

**Comment by chyzwar** [comment]
nodejs process startup time of node.js on my system is around 80ms. The biggest offender is fast-glob, but it is faster than glob in runtime. I raise issue on fast-glob repo https://github.com/mrmlnc/fast-glob/issues/314

One option would to re-wrtite cli in rust. node.js asyc IO is fairly sophisticated, I am not sure if this would be much faster. 

Assuming you are ok with this approach, I will continue and add watch etc.

**Comment by kdy1** [comment]
I'm okay with the pr, but I was actually considering reimplementing CLI in rust, to support cross-module compilation to support stuff like const enum without performance degradation.

**Comment by chyzwar** [comment]
That source map issue is now fixed in core or this workaround is still needed?

I think it is a matter of priorities. I could spend some time fixing issues and implementing missing flags and options in current node.js code but cross-module compilation would not get solved.

**Comment by kdy1** [comment]
The core is not fixed yet :(
I'm now working on https://github.com/swc-project/swc/issues/846 and I'll see if I can fix https://github.com/swc-project/swc/issues/1388 with next version.

**Comment by chyzwar** [comment]
Options fixes
1. fixed missing flag for deleteDirOnStart (missing in options)
Also do not re-create outDir as successfully compilation would do this Instead

2. removed logWatchCompilation flag as it was always false and missing in options
Instead use quiet flag and follow babel cli, only show errors when --quiet

Logging.
- Calculate initial compilation time to include globbing.
- Better format time diff (12.12 ms) as there will be never an second :p

Other: 
- removed lodash
- removed source-map-convert

Performance: 

Before
```
Successfully compiled 374 files with swc.
Compilation dir1: 561.012ms
Done in 0.80s.
```

After
```
Successfully compiled: 374 files with swc (235.57ms)
Done in 0.45s.
```

**Comment by chyzwar** [comment]
@kdy1  let me know if you want to change anything.

File: tsconfig.json [hunk 1]
```diff
@@ -9,6 +9,10 @@
     "noImplicitReturns": true,
     "noFallthroughCasesInSwitch": true,
     "esModuleInterop": true,
+    "lib": [
+      "ES2019.Array",
+      "ES2020.Promise"
```
Comment (kdy1) [inline tsconfig.json:14]: Doesn't this require shim?

**Review by kdy1 — COMMENTED** [review]
Other things LGTM.

File: tsconfig.json (hunk 1 above)
Comment (chyzwar) [reply tsconfig.json:14]: no, 
Promise.allSettled is supported from node 12.9 https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled
Array.flat is supported from node 11 https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat

**Review by kdy1 — APPROVED** [review]
Thanks!

**Comment by chyzwar** [comment]
thansk, I started to look into file transpilation.

**Pull request merged**